### ISSUE TITLE
Wayland backend, fullscreen Xwayland games don't recognize mouse input on the left half of the screen 

### ISSUE THREAD

**Issue by bastardbuddha** [qtile/qtile#3675]
### The issue:

~ ❯ qtile --version
0.21.0

I don't see anything relevant to the issue in qtile.log. Nonetheless, I am attaching it here.

[qtile.log](https://github.com/qtile/qtile/files/9036523/qtile.log)

There is a mild issue with the mouse cursor (as you can see in the photos), but it's not something breaking the experience.

Here are two photos of the issue - apologies, grim was refusing to actually capture anything:

![IMG_20220704_011203](https://user-images.githubusercontent.com/23492409/177086336-6a656dfd-8a84-46f8-a9fe-03c21dffd8c8.jpg)
![IMG_20220704_011212](https://user-images.githubusercontent.com/23492409/177086342-87001f5a-08e9-47ed-be1e-f8d489d0f49d.jpg)

As you can see, Factorio (and several other games running on XWayland) have this issue. They detect keyboard input correctly, but the mouse only works on half the screen.

My set-up is as follows:
- AMD Ryzen 5800H 
- nVidia RTX 3060 mobile GPU
- Integrated AMD GPU disabled
- Arch Linux
- Two monitors - DP-2 is an external 1080p 144Hz monitor, eDP-1 is the built-in laptop monitor - 1080p 165Hz. I also have a Lenovo Explorer WMR headset which I leave plugged in - it sits at HDMI-A-1.
- I am using kanshi to set up my monitors. I have the WMR headset disabled there.
- I use an external mouse - Elecom Huge. The issue also appears when using the trackpad, so it doesn't seem like that's the root of the problem.

I'm at a loss as to how to go about further diagnosis here. It works completely fine on X11, but the Wayland backend seems to have issues.

I'll be checking in here daily, happy to provide more info! 

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by ervinpopescu** [comment]
Have you checked that this isn't an Nvidia issue?

**Comment by bastardbuddha** [comment]
I've been searching using various keywords for a few hours yesterday and have not found anything that sounds remotely similar. It hasn't occurred to me that this might be an nVidia issue, since the windows do display correctly.

As far as I am aware, I am the only person on the internet with this problem.

**Comment by m-col** [comment]
Does it persist on git master?

**Comment by bastardbuddha** [comment]
I just finished compiling it. It's late here, so I'll edit this comment with an update tomorrow (2022.07.06) once I have some time to test it and make notes and steps to reproduce.

**Comment by m-col** [comment]
To clarify I meant qtile git master, not wlroots.

**Comment by bastardbuddha** [comment]
I did compile qtile from master.

~ ❯ qtile --version
0.21.1.dev75+g6620786f

New findings:

In the newer version, When the window is full-screened, it now works properly (it didn't in the previous one).

However when in max layout, the issue persists. I did manage to narrow it down a bit - it appears that when two XWayland windows overlap, the one in the background steals focus, or causes some other kind of input lockup.

Steps to reproduce: 

1. Start qtile as normal.
2. Launch steam from terminal (I never launch it directly because  I like seeing the output)
3. Launch any XWayland game - in my case Factorio (I did most of my testing with it - it doesn't have issues with changing window sizes and resolutions)
4. Switch to max layout
5. Move mouse around and click things to see that mouse input is not detected in the area where the steam window was.

You can go into tall mode and move the steam window around, and this does affect where the mouse stops working.

Anything else I can add here to help get to the bottom of this?

**Comment by m-col** [comment]
Hmm that is unfortunate, I thought we had fixed this bug. Now that you've updated to master, do you see any errors in your log that could be related? Could you also confirm that you're on the latest release of pywlroots?

**Comment by bastardbuddha** [comment]
Nothing in qtile.log, here's the output from a session of me logging in, playing around with the bug for a few minutes, and then posting this. _There's nothing there._

[qtile.log](https://github.com/qtile/qtile/files/9059260/qtile.log)

~ ❯ sudo pacman -Qs python-pywayland
[sudo] password for phil:
local/python-pywayland 0.4.13-1
    Python bindings for the libwayland library`

It's the version in the Arch repo. I can't seem to find a newer one.

**Comment by PhoenixClank** [comment]
I had the same issue, until I realized that the mouse wasn't working in exactly the area that the Steam window is occupying (although on another group). Since Steam and the game are both running in xwayland it sorta makes sense that they're able to steal each other's mouse input, but I think they shouldn't be, at least when they're on different groups.

**Comment by 3nprob** [comment]
Also filed in #3697, #3765.

**Comment by m-col** [comment]
> As you can see, Factorio (and several other games running on XWayland) have this issue.

Is there any pattern you're seeing as to which games have the issue and which do not? If some do and some don't, can you double check that they are definitely uing XWayland? Some window toolkits may use wayland even if WAYLAND_DISPLAY isn't set in their environment, if they support it and can find the socket.

I'm struggling to reproduce this on master :/

**Comment by 3nprob** [comment]
@m-col Just curious, what do you see as the difference between this and #3697  (given that you say you can not reproduce this but the other one does for all windows). From all I can tell these two reports describe the same behavior.

**Comment by zakk4223** [comment]
Something to keep in mind: On almost every wlroots based compositor I've tried with multiple displays, there's no 'primary' XWAYLAND display set. This has always caused me issues with some games, often manifesting as the cursor being odd (offset by some constant X/Y value, constrained in weird ways, etc). If you're having weird wayland game behavior, try using xrandr to set a primary XWAYLAND display. Something like 'xrandr --output XWAYLAND0 --primary'.

**Comment by m-col** [comment]
Interesting find @zakk4223. Perhaps this is related to this XWayland limitation: https://gitlab.freedesktop.org/xorg/xserver/-/issues/899. Is the output you're setting as primary the one at the far left?

**Comment by zakk4223** [comment]
No, it's one in the middle, and no output has a negative position.
It's a three monitor setup, with the two outside monitors being 1080p and the middle (main) being 3440x1440.

I think what actually happens is that in the absence of one of the XWAYLAND displays being set primary, the games (or some underlying library, I'm not sure) try to guess which display is primary. I _think_ it prefers either the leftmost display, or the one at 0,0. I say this become sometimes the bug manifests itself as the game going fullscreen, but only at 1080p. And if the game has a resolution dropdown, it only contains valid resolutions for the 1080p display at 0,0.

I eventually 'fixed' this in sway by having an API listener that just forced the 3440x1440 monitor to primary anytime the outputs changed

**Comment by jwijenbergh** [comment]
Does #3985 help?

**Comment by tych0** [comment]
Ping, is this fixed? seems like from #3881 it might be?

**Comment by ervinpopescu** [comment]
Another ping for this, @tych0 @bajsicki @elParaguayo, safe to close?
